hpc0.qmd

```diff
@@ -1,297 +1,1300 @@
-
+:::
 
-### Solution
 
-1: 
+## quiz {.smaller}
 
-- The name of each `.txt` file in the current directory
-- On each iteration the echo command prints a different file name
+::: {.panel-tabset}
 
-2:
+### Questions
+1. If the following is run from the wildcards directory. What will be printed?
+```{.bash code-line-numbers=false}
+for f in *.txt
+> do
+> echo $f
+> cat $f > new-file.txt
+> done
+```
+2. What will the contents of `new-file.txt` be and why?
 
+### Answers
+1.
+- The command will print the name of each `.txt` file.
+- On each iteration the `echo` command prints a different file name.
+2.
 - The content of `new-file.txt` will be the same as `xyz.txt`
 - On each iteration the contents of a different file are written to `new-file.txt`, overwriting whatever was written on the previous iteration
-- This is because we used `>` to write the output of cat to `new-file.txt`
+- This is because `>` overwrites the output of cat to `new-file.txt` for each element in the sequence.
 - Using `>>` instead would result in each file's content being appended to the end of `new-file.txt`
 
 :::
 
+
+## Editing files {.smaller}
+- There are three programs available on the system that can be used for text editing:
+  - `nano`
+  - `vi`
+  - `emacs`
+- If you are unfamiliar with any of these, we recommend you try `nano`.
+- Care is needed if transferring scripts and data between Windows and Linux due to differences in line endings.
+  - `dos2unix` and `unix2dos` may be helpful to get this right.
+
+::: {.notes}
+- Data/File transfer is covered in HPC1.
+:::
+
+
+## Shell Scripts {.smaller}
+- These are simple text files.
+  - By convention the filename ends `.sh`.
+- The file should start with a 'shebang' line which tells the Linux system what command to run:
+```{.bash code-line-numbers=false}
+#!/bin/bash`
+```
+- After the 'shebang', simply enter all the commands for the script.
+- Lines starting with '`#`' will be regarded as comments
+  - Use these to make you script easier to understand.
+- To run the script it must have executable permissions.
+- A scripts saved as '`script.sh`' can be run using:
+```{.bash code-line-numbers=false}
+./script.sh
+```
+
+
+## Shell Scripts demo {.smaller}
+
+
+## Shell Script exercise {.smaller}
+- Create and run a shell script to print out the first line of each file in a directory. Test using the wildcards directory.
+- Hints:
+  - Use the `history` command to see the commands you executed.
+  - You could pipe this into tail and direct it into a file and then use the `nano` - a text editor to modify this into a script.
+  - Modify the file permissions so you can execute the file. 
+  - To execute a script called '`script.sh`' issue the following command:
+```{.bash code-line-numbers=false}
+./script.sh
+```
+
+::: {.notes}
+- Scripts you write that you commonly want to run can be put somehwere sensible and this can be added to your PATH.
+- If there is more time, perhaps look at the `find` command.
```

Comment:
`find` is a world of pain for anything more than trivial things.  I think we should have a ponder as to what other less terrifying commands would be useful additions if there's time left over.

Reply:
I've changed this to suggest other `built-in` commands.